Move to uv and pin versions

derkweijers opened the pull request:
### Summary
<!-- Provide a concise summary of the feature and its purpose. -->
Currently versions are not pinned in the project. This can lead to unexpected behavior when various people are working on the project at the same time, as versions of packages might differ based on when the project was cloned.

`uv` is gaining popularity quickly, it is extremely fast and is able to lock dependencies in a file called `uv.lock`. By locking the dependencies. This will result in a reproducable environment across various systems. It also reduces the chance of someone accidentally using a major new version of a dependency.

### Motive
<!-- Explain the reasoning behind this feature. Include details on the problem it addresses or the enhancement it provides. -->

Generally speaking, it's best practice to provide version constraints on dependencies that are used in a project. This way, you prevent accidental major version bumps and introducing irregular behavior between systems. It also makes it possible to update packages in a controlled, automatic and predictable way with tools like Dependabot and Renovate.

At the same time, this will also enable the optimization of the Dockerfile, decreasing the size and speeding up the build time.

### Implementation
<!-- Describe how the feature was implemented. Include details on the approach taken, important decisions made, and code changes. -->

`pyproject.toml` was updated with version ranges that should currently be supported by Mesa. This also generated a `uv.lock` which holds the exact versions.

### Usage Examples
<!-- Provide code snippets or examples demonstrating how to use the new feature. Highlight key scenarios where this feature will be beneficial. -->

`uv sync` -> generate a virtual environment
`uv run <some tool>` -> run a tool/script within the virtual environment
`uv add` -> add a dependency to the project
`uv add --optional <extra> <dependency>` -> add a dependency to one of the extra groups


### Additional Notes
<!-- Add any additional information that may be relevant for the reviewers, such as potential side effects, dependencies, or related work. -->

This is still compatible with `pip` for people who are not using `uv` yet.

The docs of uv are located here: https://docs.astral.sh/uv/

jackiekazil commented:
@derkweijers it seems that @hussainsultan added uv.lock in #2785, but also dependent on resolution is #2788. 
I am thinking the best way to is remove uv in this PR? Let me know your thoughts.

derkweijers commented:
Found the issue, I've introduced 2 small bugs when fixing Ruff findings in the example repo. Apologies!

Fixed in the this PR: https://github.com/projectmesa/mesa-examples/pull/278.
Once merge, I'll trigger the actions for this PR again so it can be reviewed.

EwoutH commented:
Move to `uv` is great!

What’s the maintenance/update plan/strategy for the fixed dependencies?

jackiekazil commented:
@EwoutH periodic update?
@derkweijers - do you have another suggestion?

derkweijers commented:
Apologies for the delayed response. I was traveling back home and spending some time with the family.

Personally I've been using [Dependabot](https://github.com/dependabot) over the past few years. Uv support has been slow, so for my own projects I'm migrating to [Renovate](https://docs.renovatebot.com/) which does the same thing.

Both projects will update dependencies, lockfiles, GitHub Actions. Renovate is also able to update base images that are specified in Dockerfiles.

In my opinion if this PR gets adopted/merged, setting up Dependabot or Renovate would be one of the next steps.
